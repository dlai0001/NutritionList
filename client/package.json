{
  "name": "walmart",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@apollo/client": "^3.1.2",
    "@reach/router": "^1.3.4",
    "@testing-library/jest-dom": "^4.2.4",
    "@testing-library/user-event": "^7.1.2",
    "@types/jest": "^24.0.0",
    "@types/node": "^12.0.0",
    "@types/react": "^16.9.0",
    "@types/react-dom": "^16.9.0",
    "@types/yup": "^0.29.4",
    "formik": "^2.1.5",
    "graphql": "^15.3.0",
    "graphql-tag": "^2.11.0",
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "react-scripts": "3.4.1",
    "tachyons": "^4.12.0",
    "ts-node-dev": "^1.0.0-pre.56",
    "typeface-courier-prime": "^1.1.3",
    "typescript": "~3.7.2",
    "yup": "^0.29.3"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": "react-app"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "@testing-library/react": "^9.5.0",
    "mock-apollo-client": "^0.4.0"
  }
}
