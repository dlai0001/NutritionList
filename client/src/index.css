body {
    font: 16px Courier;
}