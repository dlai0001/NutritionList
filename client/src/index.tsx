import { ApolloProvider } from '@apollo/client'
import React from 'react'
import ReactDOM from 'react-dom'
import 'tachyons/css/tachyons.css'
import 'typeface-courier-prime'
import { App } from './App'
import './index.css'
import { graphqlClient } from './libs/graphql-client'
import * as serviceWorker from './serviceWorker'


ReactDOM.render(
  <React.StrictMode>
    <ApolloProvider client={graphqlClient}>
      <App />
    </ApolloProvider>
  </React.StrictMode>,
  document.getElementById('root')
)

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: https://bit.ly/CRA-PWA
serviceWorker.unregister()
