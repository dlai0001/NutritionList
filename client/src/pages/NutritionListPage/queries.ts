import {gql, useQuery, useMutation} from '@apollo/client'


export const nutritionListQuery = gql`
    query GetNutritionList {
        nutritionList {
            dessert
            nutritionInfo {
                calories
                fat
                carb
                protein
            }
        }
    }
`
debugger;

export const useNutritionList = () => useQuery(nutritionListQuery)

export const resetMutation = gql`
    mutation ResetNutritionList {
        nutritionList {
            reset
        }
    }
`

export const useResetList = () => useMutation(resetMutation, {
    awaitRefetchQueries: true,
    refetchQueries: ['GetNutritionList']
})

export const removeMutation = gql`
    mutation DeleteNutritionEntry($desserts: [String]!) {
        nutritionList {
            remove(desserts: $desserts)
        }
    }
`
export const useRemoveDessert = () => useMutation(removeMutation, {
    awaitRefetchQueries: true,
    refetchQueries: ['GetNutritionList']
})


export const addMutation = gql`
    mutation AddNutritionEntry($dessert: NutritionEntryInput!) {
        nutritionList {
            add(input: $dessert)
        }
    }
`

export const useAddDessert = () => useMutation(addMutation, {
    awaitRefetchQueries: true,
    refetchQueries: ['GetNutritionList']
})