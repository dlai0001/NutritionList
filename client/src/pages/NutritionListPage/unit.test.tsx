import { NutritionListPage } from '.'
import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import { nutritionListQuery, resetMutation } from './queries';


const originalError = console.error;

describe('NutritionListPage', () => {

    beforeAll(() => {
        console.error = (...args: string[]) => {
            if (/Warning.*not wrapped in act/.test(args[0])) {
                return
            }
            originalError.call(console, ...args)
        }
    })

    afterAll(() => {
        console.error = originalError
    })

    it('should reset list', async () => {
        let resetCalled = false

        const mockNultritionListQuery = {
            request: {
                query: nutritionListQuery,
            },
            result: () => ({
                data: {
                    nutritionList: resetCalled ? testData : [...testData, testEntry],
                },
            }),
        }

        const mocks = [
            // Note: There is a bug in MockProvider which requires us to put the request in the mock
            // provider 2x.
            mockNultritionListQuery,
            mockNultritionListQuery,
            {
                request: {
                    query: resetMutation,
                },
                result: () => {
                    resetCalled = true

                    return {
                        data: {
                            nutritionList: {
                                reset: true,
                            },
                        },
                    }
                },
            },
        ]

        const renderTree = render(
            <MockedProvider mocks={mocks}>
                <NutritionListPage />
            </MockedProvider>
        )

        // await end of promise queue
        await new Promise(resolve => setTimeout(resolve, 0));

        // Validate preconditions
        expect(renderTree.container.innerHTML).toMatchSnapshot()
        expect(screen.getByText('Oreo')).toBeInTheDocument()
        expect(screen.getByText('Ice Cream')).toBeInTheDocument()

        // the queries can accept a regex to make your selectors more resilient to content tweaks and changes.
        fireEvent.click(screen.getByText(/Reset/i))

        await new Promise(resolve => setImmediate(resolve, 0));

        expect(screen.getByText('Oreo')).toBeInTheDocument()

    });
});

const testData = [
    {
        dessert: 'Oreo',
        nutritionInfo: {
            calories: 437,
            fat: 18,
            carb: 63,
            protein: 4,
        },
        __typename: 'NutritionEntry',
    },
    {
        dessert: 'Nougat',
        nutritionInfo: {
            calories: 360,
            fat: 19,
            carb: 9,
            protein: 37,
        },
        __typename: 'NutritionEntry',
    },
]

const testEntry = {
    dessert: 'Ice Cream',
    nutritionInfo: {
        calories: 140,
        fat: 7,
        carb: 16,
        protein: 2,
    },
    __typename: 'NutritionEntry',
}