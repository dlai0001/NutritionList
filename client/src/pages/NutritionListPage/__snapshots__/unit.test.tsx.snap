// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NutritionListPage should reset list 1`] = `"<div class=\\"fl w-100 bg-light-gray\\"><div class=\\"flex flex-row justify-between\\"><span class=\\"f1 fw6\\">Nutrition List</span><span class=\\"f6 link dim br1 ph3 pv2 mb2 dib white bg-dark-green ba b--blue bw2 fw6\\"><i class=\\"fa fa-refresh ma2\\" aria-hidden=\\"true\\"></i>Reset Data</span></div><div class=\\"bg-washed-red fw6 flex flex-row justify-between\\"><p class=\\"purple v-mid ml2\\">0 selected</p><div class=\\"w-30 flex flex-row items-end\\"><span class=\\"f6 link dim br1 ph3 pv2 mb2 dib dark-green bg-white ma2\\"><i class=\\"fa fa-plus mr2\\" aria-hidden=\\"true\\"></i>ADD NEW</span><span class=\\"f6 link br1 ph3 pv2 mb2 dib dark-green bg-white ma2 disabled\\"><i class=\\"fa fa-trash mr2\\" aria-hidden=\\"true\\"></i>DELETE</span></div></div><table class=\\"f6 w-100 mw8 center\\"><thead><tr><th class=\\"fw6 bb b--black-20 tl pb3 pr3 bg-white\\"><input type=\\"checkbox\\"></th><th class=\\"fw6 bb b--black-20 tl pb3 pr3 bg-white\\">Dessert (100g serving)</th><th class=\\"fw6 bb b--black-20 tl pb3 pr3 bg-white\\">Calories</th><th class=\\"fw6 bb b--black-20 tl pb3 pr3 bg-white\\">Fat (g)</th><th class=\\"fw6 bb b--black-20 tl pb3 pr3 bg-white\\">Carbs (g)</th><th class=\\"fw6 bb b--black-20 tl pb3 pr3 bg-white\\">Protein (g)</th></tr></thead><tbody class=\\"lh-copy\\"><tr><td class=\\"pv3 pr3 bb b--black-20\\"><input type=\\"checkbox\\"></td><td class=\\"pv3 pr3 bb b--black-20\\">Oreo</td><td class=\\"pv3 pr3 bb b--black-20\\">437</td><td class=\\"pv3 pr3 bb b--black-20\\">18</td><td class=\\"pv3 pr3 bb b--black-20\\">63</td><td class=\\"pv3 pr3 bb b--black-20\\">4</td></tr><tr><td class=\\"pv3 pr3 bb b--black-20\\"><input type=\\"checkbox\\"></td><td class=\\"pv3 pr3 bb b--black-20\\">Nougat</td><td class=\\"pv3 pr3 bb b--black-20\\">360</td><td class=\\"pv3 pr3 bb b--black-20\\">19</td><td class=\\"pv3 pr3 bb b--black-20\\">9</td><td class=\\"pv3 pr3 bb b--black-20\\">37</td></tr><tr><td class=\\"pv3 pr3 bb b--black-20\\"><input type=\\"checkbox\\"></td><td class=\\"pv3 pr3 bb b--black-20\\">Ice Cream</td><td class=\\"pv3 pr3 bb b--black-20\\">140</td><td class=\\"pv3 pr3 bb b--black-20\\">7</td><td class=\\"pv3 pr3 bb b--black-20\\">16</td><td class=\\"pv3 pr3 bb b--black-20\\">2</td></tr></tbody></table></div>"`;
